include/oc_push.h

```diff
@@ -0,0 +1,232 @@
+#endif
+#else
+#define p_dbg(...)
+#define p_wrn(...)
+#define p_err(...)
+#endif
+
+
+#ifdef __cplusplus
+extern "C"
+{
+#endif
+
+
+#define PUSHCONF_RSC_PATH "/pushconfig"
+#define PUSHCONF_RSC_TYPE "oic.r.pushconfiguration"
+#define PUSHCONF_RSC_NAME "Push Configuration"
+
+#define PUSHRECVS_RSC_PATH "/pushreceivers"
+#define PUSHRECVS_RSC_TYPE "oic.r.pushreceiver"
+#define PUSHRECVS_RSC_NAME "Push Receiver Configuration"
+
+
+/**
+ * @brief Push Proxy state
+ */
+typedef enum {
+	OC_PP_WFP,   ///< Wait For Provisioning
+	OC_PP_WFU,   ///< Wait For Update
+	OC_PP_WFR,   ///< Wait For Response
+	OC_PP_WFUM,  ///< Wait For Update Mitigation
+	OC_PP_WFRM,  ///< Wait For Response Mitigation
+	OC_PP_ERR,   ///< Error
+	OC_PP_TOUT   ///< Timeout
+} oc_pp_state_t;
+
+/**
+ * @brief structure for handling "rt": ["oic.r.notificationselector", "oic.r.pushproxy"] Resource
+ */
+typedef struct oc_ns
+{
+	struct oc_ns *next;
+	oc_resource_t *resource;      ///< used to point ["oic.r.notificationselector", "oic.r.pushproxy"] Resource managed by iotivity-lite
+	/* notificaiton selector */
+	oc_string_t phref;            ///< oic.r.notificationselector:phref (optional)
+	oc_string_array_t prt;        ///< oic.r.notificationselector:prt (optional)
+	oc_string_array_t pif;        ///< oic.r.notificationselector:pif (optional)
+	/* push proxy */
+	oc_string_t pushtarget_di;    ///< device id of target (e.g. ocf://17087f8c-13e3-4849-4258-65af2a47df63)
+	oc_endpoint_t pushtarget_ep;  ///< endpoint of pushtarget (e.g. coaps://[fe80::b1d6]:1122)
+	oc_string_t targetpath;       ///< path in pushtarget (e.g. /pushedLightSwitch)
+	oc_string_t pushqif;          ///< oic.r.pushproxy:pushqif (deprecated)
+	oc_string_array_t sourcert;   ///< oic.r.pushproxy:sourcert
+	oc_string_t state;            ///< oic.r.pushproxy:state
+	void *user_data;              ///< used to point updated pushable Resource
+} oc_ns_t;
+
+
+/**
+ * @brief structure for member of "oic.r.pushreceiver:receivers" object array
+ */
+typedef struct oc_recv
+{
+	struct oc_recv *next;
+	oc_string_t receiveruri;  ///< oic.r.pushreceiver:receivers:receiveruri
+	oc_string_array_t rts;    ///< oic.r.pushreceiver:receivers:rts
+} oc_recv_t;
+
+
+/**
+ * @brief structure for handling for Push Receiver Resource
+ */
+typedef struct oc_recvs
+{
+	struct oc_recvs *next;
+	oc_resource_t *resource;    ///< used to point ["oic.r.pushreceiver"] Resource managed by iotivity-lite
+	OC_LIST_STRUCT(receivers);  ///< oic.r.pushreceiver:receivers object array
+} oc_recvs_t;
+
+
+
+/**
+ * @brief object used to store Resource pushed to "oic.r.pshreceiver:receivers[i].receiveruri"
+ */
+typedef struct oc_pushd_rsc_rep
+{
+	struct oc_pushd_rsc_rep *next;
+	oc_resource_t *resource;        ///< used to point any pushed Resource managed by iotivity-lite
+	oc_rep_t *rep;                  ///< payload of pushed Resource
+} oc_pushd_rsc_rep_t;
+
+
+OC_PROCESS_NAME(oc_push_process);
+
+extern char *pp_state_strs[];
```

Comment:
In public headers we should only expose what is necessary. I think this and `cli_status_strs`, `cli_statusstr`, `pp_statestr`, `pp_update_state`, `oc_init_string` are all only used in `oc_push.c`, so all these definitions should be moved there. Or into a `oc_push_internal.h` header if they are reused in some other internal implementation file.

Reply:
fixed